Maple2.Server.Game/Manager/ShopManager.cs

```diff
@@ -0,0 +1,609 @@
+                ShopId = shop.Id,
+            };
+
+            if (shop.RestockData.PersistantInventory) {
+                accountShopData[shop.Id] = data;
+            } else {
+                characterShopData[shop.Id] = data;
+            }
+        }
+
+        shop.RestockTime = data.RestockTime;
+
+        // We do not save shops with minute intervals to database
+        if (shop.RestockData!.Interval != ShopRestockInterval.Minute) {
+            using GameStorage.Request db = session.GameStorage.Context();
+            long ownerId = shop.RestockData.PersistantInventory ? session.AccountId : session.CharacterId;
+            data = db.CreateCharacterShopData(ownerId, data);
+            if (data == null) {
+                throw new InvalidOperationException($"Failed to create shop data {shop.Id} for character {session.CharacterId}");
+            }
+        }
+
+        return data;
+    }
+
+    private IDictionary<int, CharacterShopItemData> CreateShopItemData(Shop shop) {
+        Dictionary<int, CharacterShopItemData> itemData = new Dictionary<int, CharacterShopItemData>();
+        using GameStorage.Request db = session.GameStorage.Context();
+        long ownerId = shop.RestockData!.PersistantInventory ? session.AccountId : session.CharacterId;
+        foreach ((int id, ShopItem item) in shop.Items) {
+            var data = new CharacterShopItemData {
+                ShopId = shop.Id,
+                ShopItemId = id,
+                Item = item.Item,
+            };
+            if (shop.RestockData.Interval != ShopRestockInterval.Minute) {
+                data = db.CreateCharacterShopItemData(ownerId, data);
+                if (data == null) {
+                    continue;
+                }
+            }
+            itemData[id] = data;
+        }
+
+        if (shop.RestockData!.PersistantInventory) {
+            accountShopItemData[shop.Id] = itemData;
+            return accountShopItemData[shop.Id];
+        }
+        characterShopItemData[shop.Id] = itemData;
+        return characterShopItemData[shop.Id];
+    }
+
+    public void InstantRestock() {
+        if (activeShop?.RestockData == null || activeShop.RestockData.DisableInstantRestock) {
+            return;
+        }
+
+        int cost = activeShop.RestockData.Cost;
+        ShopCurrencyType currencyType = activeShop.RestockData.CurrencyType;
+        if (activeShop.RestockData.EnableCostMultiplier) {
+            (ShopCurrencyType CurrencyType, int Cost) multiplierCost = Core.Formulas.Shop.ExcessRestockCost(activeShop.RestockData);
+            cost = multiplierCost.Cost;
+            currencyType = multiplierCost.CurrencyType;
+        }
+        if (!Pay(new ShopCost {
+                Amount = cost,
+                Type = currencyType,
+                ItemId = 0,
+                SaleAmount = 0,
+            }, activeShop.RestockData.Cost)) {
+            return;
+        }
+
+        Shop? shop = session.FindShop(activeShop.Id);
```

Comment:
we should probably load this shop first so we don't risk failure after user has already paid.